pkg/machinery/role/role.go

```diff
@@ -0,0 +1,90 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package role
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+// Role represents Talos user role.
+// Its string value is used everywhere: as the the Organization value of Talos client certificate,
+// as the value of talosctl flag, etc.
+type Role string
+
+const (
+	// Admin defines Talos role for admins.
+	Admin = Role("os:admin")
+
+	// Reader defines Talos role for readers who can access read-only APIs that do not expose secrets.
+	Reader = Role("os:reader")
+
+	// Impersonator defines internal Talos role for impersonating another user (and their role).
+	Impersonator = Role("os:impersonator")
+)
+
+// Set represents a set of roles.
+type Set map[Role]struct{}
+
+// all roles, including internal ones.
+var all = MakeSet(Admin, Reader, Impersonator)
+
+// MakeSet makes a set of roles from constants.
+// Use Parse in other cases.
+func MakeSet(roles ...Role) Set {
+	res := make(Set, len(roles))
+	for _, r := range roles {
+		res[r] = struct{}{}
+	}
+
+	return res
+}
+
+// Parse parses a set of roles.
+// Returned set is always non-nil and contains all parsed roles, even if some erroneous roles were present.
+func Parse(strings []string) (Set, error) {
+	res := make(Set)
+
+	var err *multierror.Error
+
+	for _, r := range strings {
+		role := Role(r)
+		if _, ok := all[role]; !ok {
+			err = multierror.Append(err, fmt.Errorf("unexpected role %q", r))
```

Comment:
just a note: it might make sense not to fail on unknown roles, the reason is compatibility with future releases of Talos

say, in 0.12 Talos we add a new role, and someone is using client certificate with a new role to access a cluster with mixed versions (0.11/0.12) in the process of being upgraded. I don't think keeping unknown roles would be any problem for us?

Reply:
That returned error is logged and ignored by the server code. The only place where this error is fatal is in `talosctl gen csr --roles` valdiation.

Reply:
but the code seems to ignore such roles, so it won't pass them through the apid impersonator for example (e.g. control plane node on 0.11 and worker on 0.12)